design/view-change.md

```diff
@@ -0,0 +1,153 @@
+  _VIEW-CHANGE_ message could be uncooperative and don't send it to some
+  nodes. In that case since primary has a view-change certificate it is
+  guaranteed that it is possible to get required _VIEW-CHANGE_ from primary
+  and _f_ matching _VIEW-CHANGE-ACKS_ from non-faulty nodes. Most papers from
+  original authors don't describe exact protocol for this case, but in
+  [Miguel Castro master thesis](https://www.microsoft.com/en-us/research/wp-content/uploads/2017/01/thesis-mcastro.pdf) 
+  there is section 5.2 on solving problem of eventually delivering all 
+  messages even if underlying network can randomly drop them. Their proposed 
+  solution is to periodically broadcast status messages, so other nodes can 
+  learn about missing messages and retransmit them.
+  
+- After obtaining (or generating in case of new primary) correct _NEW-VIEW_
+  message each replica sets stable checkpoint to _(cn, cd)_ from _X_ and
+  records all requests in _X_ as pre-prepared. Backup replicas also broadcast
+  _PREPARE_ messages for these requests. After that processing goes on in
+  normal state. There are some caveats however.
+  
+  First one (also mentioned by authors of PBFT) is that it can happen so 
+  that some replica won't have checkpoint _(cn, cd)_ and it would require 
+  state transfer protocol to get to correct state. This could be done 
+  independent from normal request processing (they can still get prepare and 
+  commit certificates), which will be executed when state is caught up.
+  There is already similar protocol implemented in indy plenum called 
+  catch up, hopefully it could be used as is in this situation.
+  
+  Another problem is that we need somehow to skip execution of requests that
+  were ordered in previous view but view-change protocol started their
+  processing again. Authors of PBFT rely on uniqueness of ppSeqNo of requests,
+  so if node gets commit certificate for some request that it already executed
+  it just skips it. However indy plenum implementation currently resets 
+  ppSeqNo counter upon entering new view, so it cannot be relied upon.
+  One possible workaround is to check if request was already ordered using
+  ledger, but this is _O(N)_ operation with yet undetermined bounds on _N_.
+  
+  Even if we come up with some acceptable workaround and still keep ppSeqNo 
+  reset logic there is one more problem - PBFT view-change protocol allows 
+  multiple consecutive view changes, keeps history of potentially ordered 
+  requests in different views and algorithm it uses to select requests to 
+  get into new view relies on ppSeqNo constantly growing.
+  
+  Given above two paragraphs it should be seriously considered that we stop
+  resetting ppSeqNo when entering new view, otherwise it will take probably
+  unacceptable amount of time to prove that modified algorithm is correct,
+  or go on without such proof, which will seriously undermine confidence in 
+  it.
```

Comment:
@ashcherbakov 
1. We can keep uncommited batches as long as new batches selected into new view have same digests (which is not always true), otherwise we'll still need to reapply them. Given that we perform dynamic validation during pre-prepare this could become a problem, since we'll need to rerun it and it might fail, so further analysis is needed of what should be done in this case. I see several options here:
- (best one for now) prove that if we don't have more than _f_ malicious nodes then dynamic validation cannot fail. In this case we can still run it, but if it fails we should just stop processing, probably try to start a view change and log this event with error or critical level
- if we prove instead that this can happen even if no more than _f_ nodes are malicious then we should analyse if we can drop such transaction (and all later ones, since it and all later ones surely weren't ordered on normal nodes)

Also, authors of PBFT papers don't rely on dynamic validation during pre-prepare as they solve abstract problem of making sure that all replicas of state machine receive same events in same order. If we also abstract state from BFT protocol we can just move dynamic validation to execution step and treating failures as yet another transition (which doesn't actually change ledger). Also this approach can potentially increase throughput because we'll no longer need to wait for all previous requests during pre-prepare phase before continuing ordering.

2. I think we could just keep syncing only master instance during view change as we do it now, the only change is that backup instances should be set to (viewNo+1, ppSeqNo) instead of (viewNo+1, 0).